File: services/webhook-service/webhook_service/presentation/http/contract/envelope.go
package contract

import (
	"encoding/json"
	"net/http"
	"time"
)

// Envelope is the success shape every endpoint answers in.
type Envelope struct {
	Data  any            `json:"data,omitempty"`
	Error *ErrorBody     `json:"error,omitempty"`
	Meta  map[string]any `json:"meta"`
}

// ErrorBody is the error shape every endpoint fails in.
type ErrorBody struct {
	Code    string        `json:"code"`
	Message string        `json:"message"`
	Details []ErrorDetail `json:"details,omitempty"`
}

// ErrorDetail names one field a failure refused.
type ErrorDetail struct {
	Field   string `json:"field"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

const correlationHeader = "X-Correlation-ID"

// WriteOK writes a 200 in the success envelope.
func WriteOK(writer http.ResponseWriter, data any, meta map[string]any) {
	if meta == nil {
		meta = map[string]any{}
	}

	writeJSON(writer, http.StatusOK, Envelope{Data: data, Meta: meta})
}

// WriteError writes a failure in the error envelope, correlation id included.
func WriteError(
	writer http.ResponseWriter,
	request *http.Request,
	status int,
	code string,
	message string,
	details ...ErrorDetail,
) {
	writeJSON(writer, status, Envelope{
		Error: &ErrorBody{Code: code, Message: message, Details: details},
		Meta: map[string]any{
			"request_id": request.Header.Get(correlationHeader),
			"timestamp":  time.Now().UTC().Format(IsoLayout),
		},
	})
}

func writeJSON(writer http.ResponseWriter, status int, body Envelope) {
	writer.Header().Set("Content-Type", "application/json")
	writer.WriteHeader(status)

	if encodeErr := json.NewEncoder(writer).Encode(body); encodeErr != nil {
		logResponseEncodeFailed(encodeErr)
	}
}

File: services/ai-service/service_core/shared/http_contract/envelope.py
from datetime import UTC, datetime
from typing import Any

from fastapi import Request
from fastapi.responses import JSONResponse

from .config import HeaderName
from .exceptions import ApiError


def ok(data: Any, meta: dict[str, Any] | None = None) -> dict[str, Any]:
    return {"data": data, "meta": meta if meta is not None else {}}


def error_response(request: Request, failure: ApiError) -> JSONResponse:
    error: dict[str, Any] = {
        "code": str(failure.code),
        "message": failure.message,
    }

    if failure.details:
        error["details"] = [detail.as_dict() for detail in failure.details]

    return JSONResponse(
        status_code=failure.code.status_code,
        content={
            "error": error,
            "meta": {
                "request_id": request.headers.get(HeaderName.CORRELATION),
                "timestamp": datetime.now(UTC).isoformat(),
            },
        },
    )
